File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        c1 = Counter(s1)
        c2 = Counter(s2[:len(s1):])
        if c1 == c2:
            return True
        for i in range(1, len(s2) + 1 - len(s1)):
            c2[s2[i-1]] -= 1
            if c2[s2[i-1]] <= 0:
                del c2[s2[i-1]]
            if s2[i-1+len(s1)] in c2.keys():
                c2[s2[i-1+len(s1)]] += 1
            else:
                c2[s2[i-1+len(s1)]] = 1
            if c1 == c2:
                return True
        return False

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s2.length() < s1.length())
            return false;
        
        array<int, 26> slider = {0}, letters = {0};
        
        for(int i = 0; i < s1.length(); ++i) {
            ++letters[s1[i] - 'a'];
            ++slider[s2[i] - 'a'];
        }
        if(slider == letters)
            return true;
        
        for(int i = 1; i <= s2.length() - s1.length(); ++i) {
            --slider[s2[i-1] - 'a'];
            ++slider[s2[i+s1.length()-1] - 'a'];
            if(slider == letters)
                return true;
        }
        
        return false;
    }
};